whisper-cpp: add optional FFmpeg support

gaelj opened the pull request:
Whisper-cpp has an FFmpeg build flag to support additional file formats, as detailed at https://github.com/ggml-org/whisper.cpp#ffmpeg-support-linux-only.

This change exposes the build flag in the nix pkg.

note: for whisper-server to work after FFmpeg support is enabled, it must always be started with the `--convert` flag. For this reason and to prevent breaking people's configurations, we made the FFmpg build flag opt-in.

To enable it:

```nix
environment.systemPackages = [ pkgs.whisper-cpp.override {
    withFFmpegSupport = true;
} ];
```

`whisper-cli` and `whisper-server` were successfully tested (in my case, with a llama-swap front-end), albeit on my own derivation of llama-swap (which I made before it become available in nixpkgs):

```
 curl --location "https://llama.server.home.arpa/upstream/large-v1/v1/audio/transcriptions" --form 'file=@"output.opus"' --form 'language=ja' --form 'response_format=srt'
1
00:00:00,000 --> 00:00:03,000
鶏の足は、滑らかな床について、

2
00:00:03,000 --> 00:00:07,000
黒い布団に貼り付けた。
```

In ~~my version of~~ llama-swap systemd settings, I needed to add `WorkingDirectory = "/tmp";` for the conversion, so I added it to the official derivation ~~, but I could not test it myself~~.

Edit: as discussed in this pr, modifications to llama-swap have been removed from this changeset

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

aviallon reviewed (changes requested):
LGTM for the whisper-cpp changes.
However, don't modify llama-swap in this PR.
It should be done in a follow-up PR.

gaelj commented on pkgs/by-name/wh/whisper-cpp/package.nix:
```diff
@@ -146,6 +150,9 @@ effectiveStdenv.mkDerivation (finalAttrs: {
     (cmakeBool "GGML_METAL_EMBED_LIBRARY" true)
+  ]
+  ++ optionals withFFmpegSupport [
+    (cmakeFeature "WHISPER_FFMPEG" "yes")
```
Good suggestion ! I will rebase into a single commit when I get back to a proper computer

gaelj replied:
now squashed & rebased

Aleksanaa commented on pkgs/by-name/wh/whisper-cpp/package.nix:
Wait, how is this related to metal?

gaelj committed:
whisper-cpp: add FFmpeg support

as detailed at https://github.com/ggml-org/whisper.cpp#ffmpeg-support-linux-only

note: for whisper-server to work, it must be started with `--convert` (even for natively support formats, such as wav)

gaelj commented:
@Aleksanaa @hughobrien @aviallon @jsierles thank you so much for your thorough reviews !

So as mentioned above, because I have my own spin of `llama-swap` package and module, I cannot easily check if the format conversion goes well with the official one as it is, or if, as in my case, it needs to be explicitly allowed to write in `/tmp`. At least that's how I solved this error:

```
[in#0 @ 0x56457df2f040] Error opening input: No such file or directory
Error opening input file whisper-server-20251121-014036-459268320.wav.
Error opening input files: No such file or directory
```

If (and only if) it's quick and easy for any of you, may I kindly ask you to do the check and ping me back with the outcome ? (keep in mind to run `whisper-server` with the `--convert` flag)

If not then I'll migrate my config to the official llama-swap and check it myself, but it will take a bit of time. It's no problem though, I don't mean to be abusing of your limited time.

Thanks again !

aviallon commented:
> @Aleksanaa @hughobrien @aviallon @jsierles thank you so much for your thorough reviews !
> 
> So as mentioned above, because I have my own spin of `llama-swap` package and module, I cannot easily check if the format conversion goes well with the official one as it is, or if, as in my case, it needs to be explicitly allowed to write in `/tmp`. At least that's how I solved this error:
> 
> ```
> [in#0 @ 0x56457df2f040] Error opening input: No such file or directory
> Error opening input file whisper-server-20251121-014036-459268320.wav.
> Error opening input files: No such file or directory
> ```
> 
> If (and only if) it's quick and easy for any of you, may I kindly ask you to do the check and ping me back with the outcome ? (keep in mind to run `whisper-server` with the `--convert` flag)
> 
> If not then I'll migrate my config to the official llama-swap and check it myself, but it will take a bit of time. It's no problem though, I don't mean to be abusing of your limited time.
> 
> Thanks again !

Well, here is an even better idea: add a simple check that verifies the proper working of whisper-cpp with a format that requires conversion.
